node/silkworm/trie/intermediate_hashes.cpp

```diff
@@ -274,33 +321,36 @@ evmc::bytes32 DbTrieLoader::calculate_root(PrefixSet& changed) {
                     storage_collector_.collect(std::move(e));
                 };
 
-                StorageTrieCursor storage_trie{txn_};
-                for (storage_trie.seek_to_account(key_with_inc);; storage_trie.next()) {
+                for (Cursor storage_trie{storage_trie_db_cursor, changed, key_with_inc};
+                     storage_trie.key().has_value();) {
                     if (storage_trie.can_skip_state()) {
-                        goto use_storage_trie;
+                        assert(storage_trie.hash() != nullptr);
+                        storage_hb.add_branch_node(*storage_trie.key(), *storage_trie.hash(),
+                                                   storage_trie.children_are_in_trie());
+                    }
+
+                    const std::optional<Bytes> uncovered_storage{storage_trie.first_uncovered_prefix()};
+                    if (uncovered_storage == std::nullopt) {
+                        // no more uncovered storage
+                        break;
                     }
 
-                    for (auto storage{storage_state.lower_bound_multivalue(
-                             db::to_slice(key_with_inc), db::to_slice(storage_trie.first_uncovered_prefix()),
-                             /*throw_notfound=*/false)};
-                         storage.done; storage = storage_state.to_current_next_multi(/*throw_notfound=*/false)) {
+                    storage_trie.next();
+
+                    for (auto storage{storage_state.lower_bound_multivalue(db::to_slice(key_with_inc),
+                                                                           db::to_slice(*uncovered_storage),
+                                                                           /*throw_notfound=*/false)};
+                         storage; storage = storage_state.to_current_next_multi(/*throw_notfound=*/false)) {
                         const Bytes unpacked_loc{unpack_nibbles(db::from_slice(storage.value).substr(0, kHashLength))};
-                        const ByteView value{db::from_slice(storage.value).substr(kHashLength)};
                         if (storage_trie.key().has_value() && storage_trie.key().value() < unpacked_loc) {
                             break;
                         }
```

Comment:
I personally find not much readable this way to use `for` loop on mdbx cursor.
Maybe is worth to implement an `for_each_multi` in access layer ?

Reply:
Fair enough; I've left a TODO.